flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java

```diff
@@ -614,34 +612,23 @@ private void stopTaskExecutorServices() throws Exception {
 
 		if (task != null) {
 			for (final PartitionInfo partitionInfo: partitionInfos) {
-				IntermediateDataSetID intermediateResultPartitionID = partitionInfo.getIntermediateDataSetID();
-
-				final SingleInputGate singleInputGate = task.getInputGateById(intermediateResultPartitionID);
+				// Run asynchronously because it might be blocking
+				getRpcService().execute(
+					() -> {
+						try {
+							networkEnvironment.updatePartitionInfo(partitionInfo);
+						} catch (IOException | PartitionException | InterruptedException e) {
```

Comment:
We changed some previous behavior here but seems not covered by related tests before.
The `PartitionException` would trigger cancel task before, so the task final state might be canceled. But now we fail the task directly, so the task final state might be failed. Does it need to supplement a unit test?

Reply:
I think before it would send `PartitionException` back to the `JobMaster` and which would then fail the `Execution`. This should have the same effect. Verifying whether this is guarded by a test makes sense, though.

Reply:
Yes, I agree with the same effect, only concern the unit test. If we ever had the unit test for verifying the task is in `CANCELED` state after `PartitionException`, then this change should make the previous test failure because the task would be in `FAILED` state after change.

Reply:
Does it not fail the execution currently (just on JM side)?
From what I see, `Execution.sendUpdatePartitionInfoRpcCall` reacts on failed RPC by failing itself.